index.js

```diff
@@ -6,7 +6,7 @@ var isWin32 = require('os').platform() === 'win32';
 
 var slash = '/';
 var backslash = /\\/g;
-var enclosure = /[{[].*\/.*[}\]]$/;
+var enclosure = /[{[][^/\r\n\u2028\u2029]*\/.*[}\]]$/;
```

Comment:
Would this suffice (and be a fair bit easier to read as well)?

```suggestion
var enclosure = /[{[][^/]*\/.*[}\]]$/;
```

If not, we should probably add a test case that shows it not working.

Reply:
@Trott The direct solution of this issue is the way as you suggested. But the reporter suggested this solution and we agree with it because we can't accept newlines in a filepath.

Reply:
If the decision is "we won't accept newlines in a path", then OK. But if the belief is "newlines cannot legitimately appear in a path", then I do not believe that is correct. See https://superuser.com/a/129532/81159. Or try this (which is basically the same thing as in that linked answer):

```
$ touch 'foo
quote> bar'
$ ls -lbd foo*
-rw-r--r--  1 trott  staff  0 Jun 24 07:04 foo\nbar
$ 
```

Reply:
As a onetime file system tester, I'm with @Trott here: newlines are valid path elements in many file systems.

Reply:
This regex is much improved in its performance. I did not check the context in which it is used, but if one of the "match a substring within this string" APIs is used then the performance can still be problematic.

Try for example:

```
let match = '{'.repeat(10*10*10*10*5);

var res = /[[{[][^/\r\n\u2028\u2029]*\/.*[}\]]$/.exec(match);
console.log('done matching')
console.log(res)
```

(This is slow in a Node v12 REPL I found on the web).

The slow performance is because:
1. The "match anywhere" will basically try starting from each of `n` open-braces
2. There is a reachable quantifier that also matches a `{` (the first *'d group).
3. So the regex engine will match `O(n)` of them before rejecting each time -- that makes for `O(n^2)` behavior.

So, food for thought:
1. How long are the strings that this regex processes? (In the example above, I'm using a string with 50K characters in it. For shorter strings, the performance is "not great", but perhaps not catastrophic).
2.Are there any additional constraints on the input we could add in? (e.g. enforcing a maximum length, or additional properties that the regex can check) I think there is a possibly-illegible refactoring that would do the trick, but it would be nicer if there is a more maintainable option.

Reply:
I looked at the project again and it seems like the regex [`/(^|[^\\])([{[]|\([^)]+$)/`](https://github.com/gulpjs/glob-parent/blob/3ad9597edfff30f8deb9f35e9f2554a618bd8656/index.js#L10) has the same problem for all inputs `"(".repeat(n)`.

Whatever solution is chosen for `/[[{[][^/\r\n\u2028\u2029]*\/.*[}\]]$/` should also be applied to `/(^|[^\\])([{[]|\([^)]+$)/`.

Reply:
I imagine that ideally, this wouldn't be doing regex matching at all but would instead use some module that does proper parsing for globs. Although I guess there really isn't a single globbing standard to adhere to. Anyway, if no one's looked yet, it might be worth looking at what's out there (to make sure they just don't do regexp matching under the hood).

Reply:
Another option is to use a linear-time regex engine. For example, you could add `re2` as a (somewhat heavy, alas) dependency. It supports all of the features used in these regexes.

Reply:
(Please @ me if you want further opinions)

Reply:
@RunDevelopment I think that regex has no problem.
```
% cat test.js
var performance = require('perf_hooks').performance;
var globby = /(^|[^\\])([{[]|\([^)]+$)/;
var st = performance.now();
globby.test("{" + "/".repeat(500000));
var ed = performance.now();                                                     
console.log(ed - st);
% node test.js
0.7820440530776978
%
```